### PR TITLE
Add a local storage as a cache storage, cover wrap with cache with tests

### PR DESCRIPTION
**Pull request by oleh-momot**

### CODE REVIEW

File: ilc/package.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
   "description": "",
   "main": "server/app.js",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1",
+    "test": "mocha",
```
Comment (oleh-momot) [inline ilc/package.json]: **TODO**
Add test script to our CI

File: ilc/common/wrapWithCache.js [hunk 2]
```diff
@@ -1,63 +1,55 @@
...
 
             return cacheResolutionPromise[hash];
         }
 
-        return Promise.resolve(cache[hash]);
+        return Promise.resolve(cacheStorage.get(hash));
     };
 };
 
-function hashFn(s) {
```
Comment (StyleT) [inline ilc/common/wrapWithCache.js]: I would prefer to keep it here. No one should use it outside of the cache provider as it's too weak.

File: ilc/spec/hashFn.spec.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import chai from 'chai';
+import hashFn from '../common/hashFn';
+
+describe('hashFn', () => {
```
Comment (StyleT) [inline ilc/spec/hashFn.spec.js]: I would say that these tests a little bit useless :) Let's discuss tomorrow

File: ilc/spec/wrapWithCache.spec.js [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+            chai.expect(cachedValue).to.own.include({
+                data,
+            }).and.to.have.all.keys('checkAfter', 'cachedAt');
+        });
+
+        it('should save a value into the cache', async () => {
+            await wrapedFn();
+
+            chai.expect(cacheStorage.size).to.equal(1);
```
Comment (StyleT) [inline ilc/spec/wrapWithCache.spec.js]: I don't think this is a correct test. Let's discuss

File: ilc/spec/wrapWithCache.spec.js [hunk 5]
```diff
@@ -0,0 +1,177 @@
+import chai from 'chai';
+import sinon from 'sinon';
+import wrapWithCache from '../common/wrapWithCache';
+
+describe('wrapWithCache', () => {
```
Comment (StyleT) [inline ilc/spec/wrapWithCache.spec.js]: You haven't tested concurrent data request with cold cache

File: ilc/spec/wrapWithCache.spec.js [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+            chai.expect(cacheStorage.size).to.equal(1);
+        });
+
+        describe('but fetching a data was rejected', () => {
+            beforeEach(() => {
+                fn.withArgs().returns(Promise.reject(errorThrowedByFn));
+            });
+
+            it('should not save any value into the cache', async () => {
```
Comment (StyleT) [inline ilc/spec/wrapWithCache.spec.js]: I don't think this is a correct test.

File: ilc/spec/wrapWithCache.spec.js [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+            it('should reject an error when cashe storage does not have an old data', async () => {
+                try {
+                    await wrapedFn();
+                } catch (error) {
+                    chai.expect(error).to.equal(errorThrowedByFn);
+                }
+            });
+
+            it('a logger should not notice any error', async () => {
```
Comment (StyleT) [inline ilc/spec/wrapWithCache.spec.js]: I don't think this is a correct test.

File: ilc/spec/wrapWithCache.spec.js [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+            before(() => {
+                wrapedFn = wrapWithCacheStorage(fn, cacheParams);
+            });
+
+            beforeEach(() => {
+                fn.withArgs(...fnArgs).returns(Promise.resolve(newData));
+                cacheStorage.set(cachedValueKey, oldCachedValue);
+            });
+
+            it('should return a previous cached value when a user is calling a cached function the first time after expiring data', async () => {
```
Comment (StyleT) [inline ilc/spec/wrapWithCache.spec.js]: This & the following test can be merged

File: ilc/client/localStorageCache/LocalStorageCache.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
+export default class LocalStorageCache {
```
Comment (StyleT) [inline ilc/client/localStorageCache/LocalStorageCache.js]: Place it near wrapWithCache. Let's discuss

Comment (nightnei) [inline ilc/client/localStorageCache/LocalStorageCache.js]: I propose to remove wrapper end emulation and use library like https://www.npmjs.com/package/node-localstorage

**Comment by StyleT** [comment]
How I would structure tests:

- It returns cached value
- It returns cached value while respecting func params as key
- It returns stale value while requesting new value in background
- It does only one data request in case of concurrent invocation
- It logs error if fails to request data in background
- It throws error if fails to request data & no cache available
- etc...

File: ilc/common/wrapWithCache.js (hunk 2 above)
Comment (oleh-momot) [reply ilc/common/wrapWithCache.js]: Removed changes

File: ilc/spec/hashFn.spec.js (hunk 3 above)
Comment (oleh-momot) [reply ilc/spec/hashFn.spec.js]: Removed tests

File: ilc/spec/wrapWithCache.spec.js (hunk 4 above)
Comment (oleh-momot) [reply ilc/spec/wrapWithCache.spec.js]: Updated

File: ilc/spec/wrapWithCache.spec.js (hunk 7 above)
Comment (oleh-momot) [reply ilc/spec/wrapWithCache.spec.js]: Updated

File: ilc/spec/wrapWithCache.spec.js (hunk 8 above)
Comment (oleh-momot) [reply ilc/spec/wrapWithCache.spec.js]: Merged

File: ilc/spec/wrapWithCache.spec.js (hunk 6 above)
Comment (oleh-momot) [reply ilc/spec/wrapWithCache.spec.js]: Updated

File: ilc/spec/wrapWithCache.spec.js (hunk 5 above)
Comment (oleh-momot) [reply ilc/spec/wrapWithCache.spec.js]: Updated

File: ilc/.mocharc.json [hunk 10]
```diff
@@ -0,0 +1,6 @@
+{
+    "spec": "./spec/**/*.spec.js",
+    "require": "@babel/register",
+    "timeout": 5000,
+    "exit": true
```
Comment (StyleT) [inline ilc/.mocharc.json]: I don't think we need this

File: ilc/.mocharc.json [hunk 11]
```diff
@@ -0,0 +1,6 @@
+{
+    "spec": "./spec/**/*.spec.js",
```
Comment (StyleT) [inline ilc/.mocharc.json]: What about `**/*.spec.js`? So we can place tests near code.

File: ilc/.mocharc.json (hunk 10 above)
Comment (oleh-momot) [reply ilc/.mocharc.json]: Removed

File: ilc/.mocharc.json (hunk 11 above)
Comment (oleh-momot) [reply ilc/.mocharc.json]: Moved

File: ilc/client/localStorageCache/LocalStorageCache.js (hunk 9 above)
Comment (oleh-momot) [reply ilc/client/localStorageCache/LocalStorageCache.js]: Updated

File: ilc/package.json (hunk 1 above)
Comment (oleh-momot) [reply ilc/package.json]: Added

**Pull request merged**